File: internal/core/assurance.go
package core

import (
	"context"
	"database/sql"
	"errors"
)

type ScreenAssurance struct {
	ScreenID         string           `json:"screen_id"`
	Name             string           `json:"name"`
	Market           string           `json:"market"`
	Format           string           `json:"format"`
	ReservedPlays    int64            `json:"reserved_plays"`
	SettledPlays     int64            `json:"settled_plays"`
	ReleasedPlays    int64            `json:"released_plays"`
	HeldMicros       int64            `json:"held_micros"`
	SettledMicros    int64            `json:"settled_micros"`
	AwaitingEvidence int64            `json:"awaiting_evidence"`
	Decisions        map[string]int64 `json:"decisions"`
}
type CampaignAssurance struct {
	Campaign         Campaign          `json:"campaign"`
	Screens          []ScreenAssurance `json:"screens"`
	ObservedAt       int64             `json:"observed_at"`
	Scope            string            `json:"scope"`
	BudgetConsistent bool              `json:"budget_consistent"`
	SettledPlays     int64             `json:"settled_plays"`
	HeldPlays        int64             `json:"held_plays"`
	AwaitingEvidence int64             `json:"awaiting_evidence"`
	Decisions        map[string]int64  `json:"decisions"`
}

// Assurance aggregates reservation money separately from receipt decisions:
// joining money to many receipts would multiply spend under redelivery.
func (s *Store) Assurance(ctx context.Context, tenant, campaignID string) (CampaignAssurance, error) {
	out := CampaignAssurance{ObservedAt: millis(), Screens: []ScreenAssurance{}, Decisions: map[string]int64{}, Scope: "All retained campaign reservations and linked receipts. Operational playback evidence, not audience measurement. Unknown-reservation receipts cannot be attributed to a campaign."}
	if !validID(campaignID) {
		return out, invalid("invalid_campaign_id")
	}
	opts := &sql.TxOptions{ReadOnly: true}
	if s.Postgres {
		opts.Isolation = sql.LevelRepeatableRead
	}
	tx, err := s.DB.BeginTx(ctx, opts)
	if err != nil {
		return out, err
	}
	defer tx.Rollback()
	c := &out.Campaign
	err = tx.QueryRowContext(ctx, `SELECT id,name,budget,reserved,spent FROM campaigns WHERE tenant=$1 AND id=$2`, tenant, campaignID).Scan(&c.ID, &c.Name, &c.Budget, &c.Reserved, &c.Spent)
	if errors.Is(err, sql.ErrNoRows) {
		return out, ErrNotFound
	}
	if err != nil {
		return out, err
	}
	rows, err := tx.QueryContext(ctx, `SELECT s.id,s.name,s.market,s.format,
 SUM(CASE WHEN r.state='held' THEN 1 ELSE 0 END),SUM(CASE WHEN r.state='settled' THEN 1 ELSE 0 END),SUM(CASE WHEN r.state='released' THEN 1 ELSE 0 END),
 SUM(CASE WHEN r.state='held' THEN r.cost ELSE 0 END),SUM(CASE WHEN r.state='settled' THEN r.cost ELSE 0 END),
 SUM(CASE WHEN r.state='held' AND r.expires<$3 THEN 1 ELSE 0 END)
 FROM reservations r JOIN screens s ON s.id=r.screen WHERE r.tenant=$1 AND r.campaign=$2 GROUP BY s.id,s.name,s.market,s.format ORDER BY s.market,s.id`, tenant, campaignID, out.ObservedAt)
	if err != nil {
		return out, err
	}
	index := map[string]int{}
	var held, spent int64
	for rows.Next() {
		var v ScreenAssurance
		v.Decisions = map[string]int64{}
		if err = rows.Scan(&v.ScreenID, &v.Name, &v.Market, &v.Format, &v.ReservedPlays, &v.SettledPlays, &v.ReleasedPlays, &v.HeldMicros, &v.SettledMicros, &v.AwaitingEvidence); err != nil {
			rows.Close()
			return out, err
		}
		index[v.ScreenID] = len(out.Screens)
		out.Screens = append(out.Screens, v)
		held += v.HeldMicros
		spent += v.SettledMicros
		out.SettledPlays += v.SettledPlays
		out.HeldPlays += v.ReservedPlays
		out.AwaitingEvidence += v.AwaitingEvidence
	}
	err = rows.Err()
	rows.Close()
	if err != nil {
		return out, err
	}
	reservationExpr := `json_extract(d.payload,'$.reservation_id')`
	if s.Postgres {
		reservationExpr = `(d.payload::jsonb ->> 'reservation_id')`
	}
	rows, err = tx.QueryContext(ctx, `SELECT r.screen,d.status,COUNT(*) FROM deliveries d JOIN reservations r ON r.tenant=d.tenant AND r.id=`+reservationExpr+` WHERE d.tenant=$1 AND r.campaign=$2 GROUP BY r.screen,d.status`, tenant, campaignID)
	if err != nil {
		return out, err
	}
	for rows.Next() {
		var screen, status string
		var n int64
		if err = rows.Scan(&screen, &status, &n); err != nil {
			rows.Close()
			return out, err
		}
		if i, ok := index[screen]; ok {
			out.Screens[i].Decisions[status] = n
		}
		out.Decisions[status] += n
	}
	err = rows.Err()
	rows.Close()
	if err != nil {
		return out, err
	}
	out.BudgetConsistent = held == c.Reserved && spent == c.Spent && held+spent <= c.Budget
	return out, tx.Commit()
}

File: internal/console/web/assurance.js
"use strict";
const campaignAssurance = (() => {
  let selected = "",
    report = null,
    loading = false,
    error = "",
    generation = 0;
  function recommendations(r) {
    const items = [];
    if (!r.budget_consistent)
      items.push([
        "Budget mismatch",
        "Campaign balances disagree with reservation totals. Investigate before treating this report as financially reconciled.",
        "campaigns",
      ]);
    if (r.awaiting_evidence)
      items.push([
        "Follow up on missing valid playback",
        `${num(r.awaiting_evidence)} held plays are past their play window. Valid evidence may still arrive during the receipt grace period; expired holds are released by the worker.`,
        "ledger",
      ]);
    if (r.decisions.quarantined)
      items.push([
        "Review partner evidence",
        `${num(r.decisions.quarantined)} linked receipts are quarantined. Inspect their reasons; corrected content requires a new event identity. Historical quarantines remain even after a correction settles.`,
        "recovery",
      ]);
    if (r.decisions.failed)
      items.push([
        "Recover processing failures",
        `${num(r.decisions.failed)} linked receipts exhausted dispatch retries. Resolve the operational cause before replaying.`,
        "recovery",
      ]);
    if (r.decisions.accepted)
      items.push([
        "Watch processing progress",
        `${num(r.decisions.accepted)} receipts are durable but not yet reconciled. Acceptance is not a settlement.`,
        "ledger",
      ]);
    if (!items.length)
      items.push([
        "No exceptions in this report",
        "The observed balances agree and no linked receipt exceptions or overdue held plays were found. This is an operational check, not a campaign performance forecast.",
        "campaigns",
      ]);
    return items
      .map(
        (i) =>
          `<article class="assurance-action"><div><h3>${i[0]}</h3><p>${i[1]}</p></div><a class="button small" href="#${i[2]}">Inspect ↗</a></article>`,
      )
      .join("");
  }
  function content() {
    if (loading)
      return '<div class="empty">Reading the complete campaign ledger…</div>';
    if (error)
      return `<div class="error-banner" role="alert">${esc(error)} <button class="button" data-assurance-refresh>Retry report</button></div>`;
    if (!report)
      return '<div class="empty">Choose a campaign to inspect its delivery evidence.</div>';
    const r = report,
      c = r.campaign,
      available = c.budget_micros - c.spent_micros - c.reserved_micros;
    return `<div class="assurance-verdict ${r.budget_consistent ? "" : "mismatch"}"><span>${r.budget_consistent ? "✓" : "!"}</span><div><strong>${r.budget_consistent ? "Budget reconciles to the reservation ledger" : "Budget reconciliation needs investigation"}</strong><small>One consistent database snapshot · all retained campaign history · observed ${esc(new Date(r.observed_at).toLocaleString())}</small></div></div><div class="metrics">${metric("Settled spend", money(c.spent_micros), `${num(r.settled_plays)} distinct plays settled`, "↗")}${metric("Held budget", money(c.reserved_micros), `${num(r.held_plays)} plays awaiting settlement`, "◷")}${metric("Available budget", money(available), `Of ${money(c.budget_micros)} total budget`, "◫")}${metric("Duplicate reports", num(r.decisions.duplicate), "Repeated plays did not add another charge", "⧉")}</div><section class="panel"><div class="panel-header"><div><h2>Delivery by screen</h2><p>Spend comes from reservations. Receipt counts are aggregated separately.</p></div><span class="panel-tag">FULL RETAINED HISTORY</span></div>${r.screens.length ? `<div class="table-wrap"><table><thead><tr><th>SCREEN / MARKET</th><th>SETTLED PLAYS</th><th>SETTLED SPEND</th><th>HELD BUDGET</th><th>EXCEPTIONS</th></tr></thead><tbody>${r.screens.map((s) => `<tr><td><strong>${esc(s.name)}</strong><small class="proof-id">${esc(s.market)} · ${esc(s.format)} · ${esc(s.screen_id)}</small></td><td>${num(s.settled_plays)}</td><td>${money(s.settled_micros)}</td><td>${money(s.held_micros)}</td><td>${(s.decisions.quarantined || 0) + (s.decisions.failed || 0) ? `<span class="badge quarantined">${num((s.decisions.quarantined || 0) + (s.decisions.failed || 0))} receipts</span>` : '<span class="muted">None</span>'}</td></tr>`).join("")}</tbody></table></div>` : '<div class="empty">No reservations for this campaign yet. Reserve a play in Campaigns to begin.</div>'}</section><section class="panel assurance-actions"><div class="panel-header"><div><h2>Recommended operational follow-up</h2><p>Explainable rules from observed evidence. These links open workspace-wide investigation views.</p></div></div>${recommendations(r)}</section><p class="assurance-scope">${esc(r.scope)} Screen attribution uses the reserved inventory. Playback is not a measured audience impression, store visit, conversion, or proof of campaign lift.</p>`;
  }
  function update() {
    if ($("#assurance-report")) $("#assurance-report").innerHTML = content();
    const b = $("[data-assurance-export]");
    if (b) b.disabled = !report || loading || !!error;
  }
  async function fetchReport() {
    if (!selected) return;
    const ticket = ++generation;
    loading = true;
    error = "";
    report = null;
    update();
    try {
      const data = await request(
        "/api/v1/campaigns/" + encodeURIComponent(selected) + "/assurance",
      );
      if (ticket === generation) report = data;
    } catch (e) {
      if (ticket === generation) error = e.message;
    } finally {
      if (ticket === generation) {
        loading = false;
        update();
      }
    }
  }
  function render() {
    if (!selected) selected = state.campaigns[0]?.id || "";
    if (!report && !loading && !error) queueMicrotask(fetchReport);
    return (
      heading(
        "Delivery you can stand behind.",
        "Connect campaign budgets to verified plays, screen-level delivery and operational follow-up.",
        `<button class="button" data-assurance-export ${!report || loading || error ? 'disabled' : ''}>Export report ↓</button><button class="button primary" data-assurance-refresh>Refresh report ↻</button>`,
        "CAMPAIGN ASSURANCE",
      ) +
      `<div id="campaign-assurance"><div class="assurance-selector"><label for="assurance-campaign">Campaign</label><select id="assurance-campaign">${state.campaigns.map((c) => `<option value="${esc(c.id)}" ${c.id === selected ? "selected" : ""}>${esc(c.name)}</option>`).join("")}</select><span>Snapshot report · refresh to update</span></div><div id="assurance-report">${content()}</div></div>`
    );
  }
  document.addEventListener("change", (e) => {
    if (e.target.id === "assurance-campaign") {
      selected = e.target.value;
      fetchReport();
    }
  });
  document.addEventListener("click", (e) => {
    if (e.target.closest("[data-assurance-refresh]")) fetchReport();
    if (
      e.target.closest("[data-assurance-export]") &&
      report &&
      !loading &&
      !error
    ) {
      const url = URL.createObjectURL(
        new Blob([JSON.stringify(report, null, 2)], {
          type: "application/json",
        }),
      );
      const a = document.createElement("a");
      a.href = url;
      a.download = `afterglow-${report.campaign.id}-assurance.json`;
      a.click();
      setTimeout(() => URL.revokeObjectURL(url), 1000);
    }
  });
  return { render };
})();
